### ISSUE TITLE
Unable to localize the robot  in Gazebo ROS2 using Isaac ROS Map Localization

### ISSUE THREAD

**Issue by hmpcjayaweera** [NVIDIA-ISAAC-ROS/isaac_ros_map_localization#9]
Hi,
Robot - MPO-700
Simulation environment - Gazebo + ROS2 Humble

I have follow the below steps to use the Isaac ROS Map localization for the robot in Gazebo and ROS2, however, the robot model is not shown in the rviz and the "ros2 topic echo /localization_result" not give any value. It only shows that three-axis is localized (3-axis in the correct position of the rviz)
1. I have put the .png and .yaml files in the "workspaces/isaac_ros-dev/src/isaac_ros_map_localization/isaac_ros_occupancy_grid_localizer/maps" and rename them as map.png and map.yaml (also change the image e inside the .yaml to map.png). then build and source the workspace.
2. launch the MPO-700 in gazebo and ROS2 humble
3. Then I run the steps 7, 8, 9,12, and 13 in https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/blob/main/README.md#quickstart
4. when I run the steps 7, 8 and 9, the map is loaded, but the robot is not showing is Rviz, 
5. when I run the step 12, quickly a 3-axis appear on Rviz in the correct place compared to the Gazebo environment robot place. However, the "ros2 topic echo /localization_result" not give any value. 

Please help me to solve this problem. Thanks.

**Comment by ashwinvkNV** [comment]
**can you paste the launch file you are using?**

And also paste the **screenshot of the output for ros2 topic list** when the entire pipeline is running?

If you are seeing the frame, it seems like localization is working, but the topic you are looking at might be wrong.
For example we [remap the /localization_result topic to /initialpose in the Nav2 Isaac Sim launch file](https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/blob/main/isaac_ros_occupancy_grid_localizer/launch/isaac_ros_occupancy_grid_localizer_nav2.launch.py#LL88C47-L88C58).

**Comment by hmpcjayaweera** [comment]
Rviz2 as in step 7 of https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/blob/main/README.md#quickstart.
Panels:
  - Class: rviz_common/Displays
    Help Height: 138
    Name: Displays
    Property Tree Widget:
      Expanded:
        - /Global Options1
        - /Status1
      Splitter Ratio: 0.5
    Tree Height: 275
  - Class: rviz_common/Selection
    Name: Selection
  - Class: rviz_common/Tool Properties
    Expanded:
      - /2D Goal Pose1
      - /Publish Point1
    Name: Tool Properties
    Splitter Ratio: 0.5886790156364441
  - Class: rviz_common/Views
    Expanded:
      - /Current View1
    Name: Views
    Splitter Ratio: 0.5
  - Class: rviz_common/Time
    Experimental: false
    Name: Time
    SyncMode: 0
    SyncSource: ""
Visualization Manager:
  Class: ""
  Displays:
    - Alpha: 0.5
      Cell Size: 1
      Class: rviz_default_plugins/Grid
      Color: 160; 160; 164
      Enabled: true
      Line Style:
        Line Width: 0.029999999329447746
        Value: Lines
      Name: Grid
      Normal Cell Count: 0
      Offset:
        X: 0
        Y: 0
        Z: 0
      Plane: XY
      Plane Cell Count: 10
      Reference Frame: <Fixed Frame>
      Value: true
    - Alpha: 0.699999988079071
      Class: rviz_default_plugins/Map
      Color Scheme: map
      Draw Behind: false
      Enabled: true
      Name: Map
      Topic:
        Depth: 5
        Durability Policy: Volatile
        Filter size: 10
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /map
      Update Topic:
        Depth: 5
        Durability Policy: Volatile
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /map_updates
      Use Timestamp: false
      Value: true
    - Alpha: 1
      Axes Length: 1
      Axes Radius: 0.10000000149011612
      Class: rviz_default_plugins/PoseWithCovariance
      Color: 255; 25; 0
      Covariance:
        Orientation:
          Alpha: 0.5
          Color: 255; 255; 127
          Color Style: Unique
          Frame: Local
          Offset: 1
          Scale: 1
          Value: true
        Position:
          Alpha: 0.30000001192092896
          Color: 204; 51; 204
          Scale: 1
          Value: true
        Value: true
      Enabled: true
      Head Length: 0.30000001192092896
      Head Radius: 0.10000000149011612
      Name: PoseWithCovariance
      Shaft Length: 1
      Shaft Radius: 0.05000000074505806
      Shape: Axes
      Topic:
        Depth: 5
        Durability Policy: Volatile
        Filter size: 10
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /localization_result
      Value: true
  Enabled: true
  Global Options:
    Background Color: 48; 48; 48
    Fixed Frame: map
    Frame Rate: 30
  Name: root
  Tools:
    - Class: rviz_default_plugins/Interact
      Hide Inactive Objects: true
    - Class: rviz_default_plugins/MoveCamera
    - Class: rviz_default_plugins/Select
    - Class: rviz_default_plugins/FocusCamera
    - Class: rviz_default_plugins/Measure
      Line color: 128; 128; 0
    - Class: rviz_default_plugins/SetInitialPose
      Covariance x: 0.25
      Covariance y: 0.25
      Covariance yaw: 0.06853891909122467
      Topic:
        Depth: 5
        Durability Policy: Volatile
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /initialpose
    - Class: rviz_default_plugins/SetGoal
      Topic:
        Depth: 5
        Durability Policy: Volatile
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /goal_pose
    - Class: rviz_default_plugins/PublishPoint
      Single click: true
      Topic:
        Depth: 5
        Durability Policy: Volatile
        History Policy: Keep Last
        Reliability Policy: Reliable
        Value: /clicked_point
  Transformation:
    Current:
      Class: rviz_default_plugins/TF
  Value: true
  Views:
    Current:
      Class: rviz_default_plugins/Orbit
      Distance: 61.61069869995117
      Enable Stereo Rendering:
        Stereo Eye Separation: 0.05999999865889549
        Stereo Focal Distance: 1
        Swap Stereo Eyes: false
        Value: false
      Focal Point:
        X: 21.15060806274414
        Y: 10.11668872833252
        Z: -14.3422269821167
      Focal Shape Fixed Size: true
      Focal Shape Size: 0.05000000074505806
      Invert Z Axis: false
      Name: Current View
      Near Clip Distance: 0.009999999776482582
      Pitch: 1.5697963237762451
      Target Frame: <Fixed Frame>
      Value: Orbit (rviz)
      Yaw: 4.700405597686768
    Saved: ~
Window Geometry:
  Displays:
    collapsed: true
  Height: 2272
  Hide Left Dock: true
  Hide Right Dock: false
  QMainWindow State: 000000ff00000000fd0000000400000000000001dc000007f1fc0200000008fb0000001200530065006c0065006300740069006f006e00000001e10000009b000000ab00fffffffb0000001e0054006f006f006c002000500072006f007000650072007400690065007302000001ed000001df00000185000000a3fb000000120056006900650077007300200054006f006f02000001df000002110000018500000122fb000000200054006f006f006c002000500072006f0070006500720074006900650073003203000002880000011d000002210000017afb000000100044006900730070006c0061007900730000000069000007f10000017800fffffffb0000002000730065006c0065006300740069006f006e00200062007500660066006500720200000138000000aa0000023a00000294fb00000014005700690064006500530074006500720065006f02000000e6000000d2000003ee0000030bfb0000000c004b0069006e0065006300740200000186000001060000030c00000261000000010000014f000007f1fc0200000003fb0000001e0054006f006f006c002000500072006f00700065007200740069006500730100000041000000780000000000000000fb0000000a005600690065007700730000000069000007f10000012300fffffffb0000001200530065006c0065006300740069006f006e010000025a000000b200000000000000000000000200000490000000a9fc0100000001fb0000000a00560069006500770073030000004e00000080000002e1000001970000000300000e6c00000051fc0100000002fb0000000800540069006d0065010000000000000e6c0000045300fffffffb0000000800540069006d0065010000000000000450000000000000000000000e6c000007f100000004000000040000000800000008fc0000000100000002000000010000000a0054006f006f006c00730100000000ffffffff0000000000000000
  Selection:
    collapsed: false
  Time:
    collapsed: false
  Tool Properties:
    collapsed: false
  Views:
    collapsed: false
  Width: 3692
  X: 148
  Y: 54
------------------------------------------------------------------------------------------------------------
isaac_ros_occupancy_grid_localizer_quickstart.launch.py for step 9

import os

from ament_index_python.packages import get_package_share_directory

from launch import LaunchDescription
from launch.actions import GroupAction
from launch_ros.actions import ComposableNodeContainer
from launch_ros.actions import Node
from launch_ros.descriptions import ComposableNode


def generate_launch_description():

    lifecycle_nodes = ['map_server']
    map_yaml_path_dir = os.path.join(
        get_package_share_directory('isaac_ros_occupancy_grid_localizer'), 'maps', 'map.yaml')

    occupancy_grid_localizer_node = ComposableNode(
        package='isaac_ros_occupancy_grid_localizer',
        plugin='nvidia::isaac_ros::occupancy_grid_localizer::OccupancyGridLocalizerNode',
        name='occupancy_grid_localizer',
        parameters=[map_yaml_path_dir, {
            'loc_result_frame': 'map',
            'map_yaml_path': map_yaml_path_dir,
        }])

    occupancy_grid_localizer_container = ComposableNodeContainer(
        package='rclcpp_components',
        name='occupancy_grid_localizer_container',
        namespace='',
        executable='component_container_mt',
        composable_node_descriptions=[
            occupancy_grid_localizer_node,
        ],
        output='screen'
    )

    # Seting transform between lidar_frame and base_link
    # since Isaac Sim does not set this transform
    baselink_lidar_publisher = Node(
        package='tf2_ros', executable='static_transform_publisher',
        output='screen',
        arguments=['0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 'base_link',
                   'lidar_frame'])

    load_nodes = GroupAction(
        actions=[
            Node(
                package='nav2_map_server',
                executable='map_server',
                name='map_server',
                output='screen',
                respawn_delay=2.0,
                parameters=[{'yaml_filename': map_yaml_path_dir}]),
            Node(
                package='nav2_lifecycle_manager',
                executable='lifecycle_manager',
                name='lifecycle_manager_localization',
                output='screen',
                parameters=[{'use_sim_time': False},
                            {'autostart': True},
                            {'node_names': lifecycle_nodes}])
        ]
    )

    # Create the launch description and populate
    ld = LaunchDescription([occupancy_grid_localizer_container,
                            baselink_lidar_publisher])

    # Add the actions to launch all of the localiztion nodes
    ld.add_action(load_nodes)

    return ld
------------------------------------------------------------------------------------------
besides I have used  ros2 service call trigger_grid_search_localization std_srvs/srv/Empty {} and ros2 launch isaac_ros_pointcloud_utils isaac_ros_laserscan_to_flatscan.launch.py

In addtion the launch file for neo MPO-700 (for gazebo ros2 humble)

**Comment by ashwinvkNV** [comment]
Can you send the screenshot for the output for **ros2 topic list** also.

Also can you confirm that you are running `ros2 topic echo /localization_result` before running `ros2 service call trigger_grid_search_localization std_srvs/srv/Empty {}`?

**Comment by hmpcjayaweera** [comment]
Hi, 
when i try to run the rviz2 (in step 7  - https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/blob/main/README.md#quickstart) and then run the step 9, I got this error,
[ERROR] [1686091364.713581586] [rviz2]: Vertex Program:rviz/glsl120/indexed_8bit_image.vert Fragment Program:rviz/glsl120/indexed_8bit_image.frag GLSL link result : 
active samplers with a different type refer to the same texture image unit

Previously it was working, but suddenly I got this, please can you help me.

**Comment by ashwinvkNV** [comment]
I do not believe the "[ERROR] [1686091364.713581586] [rviz2]: Vertex Program:rviz/glsl120/indexed_8bit_image.vert Fragment Program:rviz/glsl120/indexed_8bit_image.frag GLSL link result :" is related to the isaac_ros_map_localization package. To discuss that please [open an issue with rviz](https://github.com/ros2/rviz/issues)

**Comment by hmpcjayaweera** [comment]
Please can I know how to make the map (.yaml and .png) files. Can I make it using SLAM Toolbox (slam tool box gives .yaml and pgm, can i use any converter to convert this .pgm to png and use for Isaac ROS map localization package ).

**Comment by ashwinvkNV** [comment]
yes you can use the map generated using SLAM Toolbox. As you said, convert the .pgm to .png and use it with isaac_ros_map_localization.

**Comment by hmpcjayaweera** [comment]
still not, will let you soon

**Comment by hmpcjayaweera** [comment]
![Screenshot from 2023-06-08 12-50-11](https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/assets/48442332/c10573e9-a795-43b1-8864-9da13fd2c
![Screenshot from 2023-06-08 12-49-23](https://github.com/N
![Screenshot from 2023-06-08 12-52-22](https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/assets/48442332/305fb5ac-4fce-4877-ae74-91d1b39f21a1)
VIDIA-ISAAC-ROS/isaac_ros_map_localization/assets/48442332/beb42a86-ceda-4f3e-8516-666c3d53e446)
c17)
the ros2 topic echo /localization_result is not publishing any value

**Comment by hmpcjayaweera** [comment]
For the ros2 topic echo /flatscan, I got "The message type 'isaac_ros_pointcloud_interfaces/msg/FlatScan' is invalid", Is this possible. Thank you

**Comment by hmpcjayaweera** [comment]
![Screenshot from 2023-06-08 13-12-40](https://github.com/NVIDIA-ISAAC-ROS/isaac_ros_map_localization/assets/48442332/7956ef7e-3a96-42d2-a793-a1fbfee73d3f)
Once I Trigger the localization using a command line service call :ros2 service call trigger_grid_search_localization std_srvs/srv/Empty {}, the three axis appear as attached photo (seems robot in the same position of the map in Gazebo). However, when I check the base_link, it is in the (0,0) of the map.
